File: solutions/2025-11-20/solution.py
# Minimum Window Substring
# Difficulty: Hard
# URL: https://leetcode.com/problems/minimum-window-substring/

def minimumWindowSubstring(s: str, t: str) -> str:
  if not s or not t:
    return ''
  t_count = Counter(t)
  required = len(t_count)
  left = 0
  min_len = float('inf')
  min_window = ''
  formed = 0
  window_counts = {}

  for right in range(len(s)):
    character = s[right]
    window_counts[character] = window_counts.get(character, 0) + 1

    if character in t_count and window_counts[character] == t_count[character]:
      formed += 1

    while left <= right and formed == required:
      character = s[left]
      if right - left + 1 < min_len:
        min_len = right - left + 1
        min_window = s[left:right + 1]
      window_counts[character] -= 1
      if character in t_count and window_counts[character] < t_count[character]:
        formed -= 1
      left += 1

  return min_window

File: solutions/2025-11-20/solution.java
// Minimum Window Substring
// Difficulty: Hard
// URL: https://leetcode.com/problems/minimum-window-substring/

public String minimumWindowSubstring(String s, String t) {
  if (s == null || t == null || s.length() == 0 || t.length() == 0) {
    return "";
  }
  Map<Character, Integer> tCount = new HashMap<>();
  for (char c : t.toCharArray()) {
    tCount.put(c, tCount.getOrDefault(c, 0) + 1);
  }
  int required = tCount.size();
  int left = 0;
  int minLen = Integer.MAX_VALUE;
  String minWindow = "";
  int formed = 0;
  Map<Character, Integer> windowCounts = new HashMap<>();

  for (int right = 0; right < s.length(); right++) {
    char character = s.charAt(right);
    windowCounts.put(character, windowCounts.getOrDefault(character, 0) + 1);

    if (tCount.containsKey(character) && windowCounts.get(character) == tCount.get(character)) {
      formed++;
    }

    while (left <= right && formed == required) {
      character = s.charAt(left);
      if (right - left + 1 < minLen) {
        minLen = right - left + 1;
        minWindow = s.substring(left, right + 1);
      }
      windowCounts.put(character, windowCounts.get(character) - 1);
      if (tCount.containsKey(character) && windowCounts.get(character) < tCount.get(character)) {
        formed--;
      }
      left++;
    }

  }
  return minWindow;
}